libcontainer/cgroups/systemd/v1.go

```diff
@@ -340,14 +345,21 @@ func (m *legacyManager) Set(r *configs.Resources) error {
 		return err
 	}
 
-	// Figure out the current freezer state, so we can revert to it after we
-	// temporarily freeze the container.
-	targetFreezerState, err := m.GetFreezerState()
+	// Figure out if cgroup needs to be frozen before applying systemd properties
+	// and thawed after, while trying to avoid unnecessary state changes.
+	freezerState, err := m.GetFreezerState()
 	if err != nil {
 		return err
 	}
-	if targetFreezerState == configs.Undefined {
-		targetFreezerState = configs.Thawed
+	needsFreeze := true
+	if freezerState == configs.Frozen {
```

Comment:
This could result in a race between the freezing of the current cgroup and one of its ancestors.

If one ancestor is FROZEN, this will not freeze the control group. If the parent is thowed before systemd does its device update sequence, things will not work as expected.

Reply:
This does also not fix the difference between cgroup v1 and v2.

If the parent control group is frozen, and the control group itself is not, `GetFreezerState` will return `FROZEN` on v1 and `THAWED` on v2. I don't think the naming or definition is important, other than the fact that it works the same way on both.

Reply:
If that make sense to you @kolyshkin and @cyphar?

Reply:
But overall, I agree that always thawing the cgroup is a valid way to do it. If people want to freeze it.

However, doing:

```
runc pause container
runc update container
# Contianer will now still be paused
runc resume container
```

now works as expected, but with your pr it will just resume the container on update (or)?

Reply:
> This could result in a race between the freezing of the current cgroup and one of its ancestors.

You are right, and I have anticipated that -- this is noted in the commit and PR description.